File: tests/cuda/julia/test_signal_spectral.jl
# Signal spectral CUDA binding tests
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_signal_spectral.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

# Try GPU
try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

function make_sine(freq::Float64, fs::Float64, n::Int)
    t = [sin(2 * pi * freq * i / fs) for i in 0:n-1]
    cpu_arr = Insight.from_data(t)
    return Insight.to_array(cpu_arr, Insight.GPUPlace(0))
end

# ============================================================================
# Spectral Analysis
# ============================================================================
println("=== Signal Spectral ===")

# welch
x = make_sine(10.0, 256.0, 1024)
f, pxx = Insight.welch_jl(x, fs=256.0, nperseg=256)
check("welch", Insight.numel(f) > 0 && Insight.numel(pxx) > 0)

# periodogram
x = make_sine(10.0, 256.0, 512)
f, pxx = Insight.periodogram_jl(x, fs=256.0)
check("periodogram", Insight.numel(f) > 0 && Insight.numel(pxx) > 0)

# csd
x = make_sine(10.0, 256.0, 1024)
y = make_sine(10.0, 256.0, 1024)
result = Insight.csd(x, y, fs=256.0, nperseg=256)
check("csd", Insight.numel(result.f) > 0 && Insight.numel(result.Pxx) > 0)

# coherence
x = make_sine(10.0, 256.0, 1024)
y = make_sine(10.0, 256.0, 1024)
result = Insight.coherence(x, y, fs=256.0, nperseg=256)
check("coherence", Insight.numel(result.f) > 0 && Insight.numel(result.Pxx) > 0)

# spectrogram
x = make_sine(10.0, 256.0, 1024)
result = Insight.spectrogram(x, fs=256.0, nperseg=256)
check("spectrogram", Insight.numel(result.f) > 0 && Insight.numel(result.t) > 0 && Insight.numel(result.Sxx) > 0)

# stft
x = make_sine(10.0, 256.0, 1024)
result = Insight.stft(x, fs=256.0, nperseg=256)
check("stft", Insight.numel(result.f) > 0 && Insight.numel(result.t) > 0 && Insight.numel(result.Sxx) > 0)

# vectorstrength
events = Insight.from_data([0.0, 1.0, 2.0, 3.0])
events_gpu = Insight.to_array(events, Insight.GPUPlace(0))
result = Insight.vectorstrength(events_gpu, 1.0)
check("vectorstrength", result.strength >= 0.0 && result.phase >= 0.0)

# welch with custom params
x = make_sine(10.0, 256.0, 2048)
f, pxx = Insight.welch_jl(x, fs=256.0, nperseg=512, noverlap=256, scaling="spectrum")
check("welch_custom", Insight.numel(f) > 0 && Insight.numel(pxx) > 0)

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end

File: tests/cuda/lua/test_signal_spectral.lua
-- Signal spectral CUDA binding tests
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
--   ~/.luarocks/bin/busted tests/cuda/lua/test_signal_spectral.lua

local ok_ins, ins = pcall(require, "_insight")
if not ok_ins then
  print("SKIP: _insight module not available")
  os.exit(0)
end

local ok_gpu = pcall(function()
  ins.init()
end)
if not ok_gpu then
  print("SKIP: GPU backend not available")
  os.exit(0)
end

ins.set_device(ins.GPUPlace(0))

describe("Signal Spectral CUDA Tests", function()
  -- Helper: build a sine wave of length n at freq Hz, sampled at fs Hz
  local function make_sine(freq, fs, n)
    local t = {}
    for i = 0, n - 1 do
      t[i + 1] = math.sin(2 * math.pi * freq * i / fs)
    end
    return ins.from_table(t):to(ins.GPUPlace(0))
  end

  it("welch", function()
    local x = make_sine(10, 256, 1024)
    local result = ins.signal.welch(x, 256.0, "hann", 256, 128, 256, "density")
    assert.is_not_nil(result)
  end)

  it("periodogram", function()
    local x = make_sine(10, 256, 512)
    local result = ins.signal.periodogram(x, 256.0, "hann", 256, "density")
    assert.is_not_nil(result)
  end)

  it("csd", function()
    local x = make_sine(10, 256, 1024)
    local y = make_sine(10, 256, 1024)
    local result = ins.signal.csd(x, y, 256.0, "hann", 256, 128, 256)
    assert.is_not_nil(result)
  end)

  it("coherence", function()
    local x = make_sine(10, 256, 1024)
    local y = make_sine(10, 256, 1024)
    local result = ins.signal.coherence(x, y, 256.0, "hann", 256, 128, 256)
    assert.is_not_nil(result)
  end)

  it("spectrogram", function()
    local x = make_sine(10, 256, 1024)
    local result = ins.signal.spectrogram(x, 256.0, "hann", 256, 128, 256)
    assert.is_not_nil(result)
  end)

  it("stft", function()
    local x = make_sine(10, 256, 1024)
    local result = ins.signal.stft(x, 256.0, "hann", 256, 128, 256)
    assert.is_not_nil(result)
  end)

  it("vectorstrength", function()
    local events = ins.from_table({ 0.0, 1.0, 2.0, 3.0 }):to(ins.GPUPlace(0))
    local result = ins.signal.vectorstrength(events, 1.0, 1.0)
    assert.is_not_nil(result)
  end)
end)
